File: js/problems/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  let max = 0;
  let sum = 0;

  for (let i = 0; i < nums.length; i++) {
    sum = 0

    for (let j = i; j < nums.length; j++) {
      sum += nums[j];

      if (sum > max) max = sum;
    }
  }

  return max;
};

console.log(maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4])); // 6
console.log(maxSubArray([5, 4, -1, 7, 8])); // 23

File: .leetcode/53.maximum-subarray.cpp
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution {
 public:
  int maxSubArray(vector<int>& nums) {
    /*
      Approach 1: Two Pointers (slow and fast)
      time complexity: O(n)
      space complexity: O(1)
    */

    int currentSum = nums[0];
    int maxSum = nums[0];

    for (int i = 1; i < nums.size(); ++i) {
      currentSum = max(nums[i], currentSum + nums[i]);
      maxSum = max(maxSum, currentSum);
    }

    return maxSum;
  }
};
// @lc code=end
